arroyo-worker/src/connectors/kinesis/source/mod.rs

```diff
@@ -0,0 +1,552 @@
+        shard_iterator: String,
+    ) -> Result<AsyncResult> {
+        let mut retries = 0;
+        loop {
+            let get_records_call = kinesis_client.get_records().shard_iterator(&shard_iterator);
+            match get_records_call.send().await {
+                Ok(result) => return Ok(AsyncResult::GetRecords(result)),
+                Err(error) => match &error {
+                    SdkError::ServiceError { err, raw: _ } => {
+                        if err.is_expired_iterator_exception() {
+                            warn!("Expired iterator exception, requesting new iterator");
+                            return Ok(AsyncResult::NeedNewIterator);
+                        }
+                        if err.is_kms_throttling_exception()
+                            || err.is_provisioned_throughput_exceeded_exception()
+                        {
+                            // TODO: make retry behavior configurable
+                            if retries == 5 {
+                                bail!("failed after {} retries", retries);
+                            }
+                            retries += 1;
+                            tokio::time::sleep(Duration::from_millis(200 * (1 << retries))).await;
+                        } else {
+                            return Err(anyhow!(error));
+                        }
+                    }
+                    _ => return Err(anyhow!(error)),
+                },
+            }
+        }
+    }
+
+    async fn handle_get_records(
+        &mut self,
+        shard_id: String,
+        get_records: GetRecordsOutput,
+        ctx: &mut Context<(), T>,
+    ) -> Result<Option<BoxedFuture<AsyncNamedResult<AsyncResult>>>, UserError> {
+        let last_sequence_number = get_records.records().and_then(|records| {
+            records
+                .last()
+                .map(|record| record.sequence_number().unwrap().to_owned())
+        });
+
+        let next_shard_iterator = self.process_records(get_records, ctx).await?;
+        let shard_state = self.shards.get_mut(&shard_id).unwrap();
+
+        if let Some(last_sequence_number) = last_sequence_number {
+            shard_state.offset = KinesisOffset::SequenceNumber(last_sequence_number.to_string());
+        }
+
+        match next_shard_iterator {
+            Some(shard_iterator_id) => Ok(Some(self.next_read_future(shard_id, shard_iterator_id))),
+            None => {
+                shard_state.closed = true;
+                Ok(None)
+            }
+        }
+    }
+    async fn handle_need_new_iterator(
+        &mut self,
+        shard_id: String,
+    ) -> Result<Option<BoxedFuture<AsyncNamedResult<AsyncResult>>>, UserError> {
+        let shard_state = self.shards.get_mut(&shard_id).unwrap();
+        Ok(Some(shard_state.get_update_shard_iterator_future(
+            self.kinesis_client.as_ref().unwrap(),
+        )))
+    }
+
+    fn get_shard_iterator_call(&self) -> GetShardIterator {
```

Comment:
It looks like this is now unused and can be deleted (not sure why it's not generating a warning for unused code...)